cve_bin_tool/version_compare.py

```diff
@@ -0,0 +1,194 @@
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+import re
+
+"""
+A class for comparing arbitrary versions of products.
+
+Splits versions up using common whitespace delimiters and also splits out letters
+so that things like openSSL's 1.1.1y type of version will work too.
+
+This may need some additional smarts for stuff like "rc" or "beta" and potentially for
+things like distro versioning.  I don't know yet.
+"""
+
+
+class CannotParseVersionException(Exception):
+    """
+    Thrown if the version doesn't comply with our expectations
+    """
+
+
+def parse_version(version_string: str):
+    """
+    Splits a version string into an array for comparison.
+    This includes dealing with some letters.
+
+    e.g. 1.1.1a would become [1, 1, 1, a]
+    """
+    versionString = version_string
+    versionArray = []
+
+    # convert - and _ to be treated like . below
+    # we could switch to a re split but it seems to leave blanks so this is less hassle
+    versionString = versionString.replace("-", ".")
+    versionString = versionString.replace("_", ".")
+
+    # Attempt a split
+    split_version = versionString.split(".")
+
+    # if the whole string was numeric then we're done and you can move on
+    if versionString.isnumeric():
+        versionArray = split_version
+        return versionArray
+
+    # Go through and split up anything like 6a in to 6 and a
+    number_letter = re.compile("([0-9]+)([a-zA-Z]+)")
+    letter_number = re.compile("([a-zA-Z]+)([0-9]+)")
+    for section in split_version:
+        # if it's all letters or all nubmers, just add it to the array
+        if section.isnumeric() or section.isalpha():
+            versionArray.append(section)
+
+        # if it looks like 42a split out the letters and numbers
+        elif re.match(number_letter, section):
+            result = re.findall(number_letter, section)
+
+            # We're expecting a result that looks like [("42", "a")] but let's verify
+            # and then add it to the array
+            if len(result) == 1 and len(result[0]) == 2:
+                versionArray.append(result[0][0])
+                versionArray.append(result[0][1])
+            else:
+                raise CannotParseVersionException(f"version string = {versionString}")
+
+        # if it looks like rc1 or dev7 we'll leave it together as it may be some kind of pre-release
+        # and we'll probably want to handle it specially in the compare.
```

Comment:
Why then not doing the same for things like 1a? It there's going to be some code for something like 'dev1', why not reusing it for '1a'? By doing that, this whole function becomes a replace of weird characters with '.' and then split.

Reply:
So, this is because versions are stupid and I'm handling two distinct cases and approaches to letters in various project's personal versions of semantic versioning:

1. Python and others have a set of things that mean "pre release" and most of them look like dev5 or rc3 or whatever..  So in this case, we want to try to compare apples to apples so that 3.4pre7 will be later than 3.4pre5 but both will be **earlier** than 3.4.  (Note this results in dev4 vs rc3 is going to default to dev being earlier because alphabetic, which is probably a bad idea but there's no way for me to know what people actually wanted in that case.)
  - e.g.  `3.4.pre7 < 3.4`
3. But then other products like openssl and libjpg use letters as... basically a number?  So we want 1.1.1z to be later than 1.1.1f but we want both of those to be **later** than 1.1.1
  - e.g. `1.1.1z > 1.1.1`

So I explicitly need to treat letters have a number after them (e,.g,. `rc4`) to be treated differently than letters that have no number after them.  I thought leaving the "pre" case grouped was easier to read in the code than doing some sort of look-back option and safer than trying to enumerate all the known strings that mean "pre" across all known software in the world (although honestly we probably could do that?  At least for the ones with reported vulns), but I'm open to suggestions on how to make it more elegant as long as that functionality remains.

Incidentally, in the original version of the code (which you can see in the diff) we converted the trailing solo letters to numbers just for those two projects.  We could go back to that and then just let the rest figure itself out, maybe?  But that's assuming that those two products are the only ones who use letters this way, which may be unlikely.  And it would leave us with weird behaviour if someone tried to compare 6.7a with 6.7.1 (they'd appear to be the same).

Reply:
Thanks! I dislike all of this (the problem, not the solution!). I would need to spend more time on the code, but would it be worth it having a precheck function or something that validates the version for the cases that you currently know and raises an exception if the format doesn't follow what you expect? That way you get an error rather that some silent corruption.